hq/markdown/snyk.md

```diff
@@ -0,0 +1,227 @@
+# Snyk
+
+## Introduction
+
+Snyk is a managed service which uses a dependency analysis tool and a vulnerability database to inform you of
+vulnerabilities in your application.
+
+## Installing
+
+Snyk requires node, and `npm`.  To install on a mac, please use:
+
+```
+brew install npm
+npm install -g snyk
+```
+
+Snyk then requires an authentication token to be set up.  Github auth is almost certainly most appropriate.
+
+```
+snyk auth
+```
+
+# Usage
+
+## Typical Usage for Applications
+
+In the build process (eg TeamCity), you can add a build step to check with the Snyk database for vulnerabilities.
+
+This information can be simply treated as reporting information, or can be used to fail a build if the return
+code is non-zero (ie at least one vulnerability found or error).
+
+## Use with Scala
+
+### Pre-requisites
+
+Snyk needs to first create a dependency tree for your project (based on the sbt dependencies), then check with the
+Snyk database.
+
+Dependency graph creation is via an sbt plugin.  Add the following line to your project plugins file or your `.sbt/plugins`
+directory:
+
+
+```
+/Users/<you>/.sbt/<sbt-version>/plugins/snyk.sbt
+or
+<projectroot>/project/plugins.sbt
+```
+
+```
+addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.9.0")
+```
+
+### Invocation
+
+It is recommended that `sbt` itself is invoked first, to confirm that the build file is good.  You
+can then invoke `snyk` as follows.  If you are working on a new snyk implementation, then it is often
+worth cutting down the on-screen noise by using `--show-vulnerable-paths=false`.
+
+```
+sbt test
+snyk test --file=build.sbt [--show-vulnerable-paths=(true|false)]
+```
+
+### Gotchas
+
+#### With SBT 1.0.0+
+
+It appears that sbt expects to invoke `dependency-tree`.  After sbt 1.0, this command appears to be `dependencyTree`.
+
+As Snyk is written in an interpreted script, this can be 'fixed' for now (if you use sbt > 1.0.0) using the following (mac) command:
+
+```
+sed -i '' 's/dependency-tree/dependencyTree/g' $(grep -wrl 'dependency-tree' /usr/local/lib/node_modules/snyk/node_modules/snyk-sbt-plugin/)
+```
+
+And to revert:
+```
+sed -i '' 's/dependencyTree/dependency-tree/g' $(grep -wrl 'dependencyTree' /usr/local/lib/node_modules/snyk/node_modules/snyk-sbt-plugin/)
+```
+
+#### With bugs in build.sbt
+
+If there is a bug in build.sbt, snyk is likely to get stuck on the `Querying vulnerabilities database...` stage.  Therefore
+it is suggested that you always invoke `sbt test` or similar first.  If  you see the `Querying vulnerabilities database...`
+step for more than ten seconds or so, you should check your `build.sbt` is good.
+
+## Use with Nodejs
+
+### Pre-requisites
+
+Ensure your node packages are up to date with the following:
+
+```
+rm node_modules
```

Comment:
You sure this is required? Could impact other stuff in the project?

Reply:
It shouldn't do any harm.  It was intended to prevent checking vulnerabilities on present dependencies, then building/packaging with newer.

Reply:
Minor: Some projects use Yarn, and have custom install commands ..they may also have a lot of dependencies so reinstalling the node modules would be something I would groan about doing.

Is there any way this could be optional, or more generic? 

```
Ensure your node packages are up to date by using the appropriate command for your project. It will probably be either: `yarn upgrade` or `npm update`
```

Would an update be enough?

Reply:
Will `rm` work? Or are we going to need to `rm -r` if node_modules is a directory?

Reply:
Fair points.  Amended.